### PR TITLE
GODRIVER-1609 Ensure a single ConnectionClosed event is emitted per connection

### PR DESCRIPTION
**Pull request by divjotarora**

### CODE REVIEW

File: x/mongo/driver/topology/pool.go [hunk 1]
```diff
@@ -83,33 +83,22 @@ func connectionExpiredFunc(v interface{}) bool {
...
-			Type:         event.ConnectionClosed,
-			Address:      c.pool.address.String(),
-			ConnectionID: c.poolID,
-			Reason:       reason,
-		})
-	}
-
-	return res
+	switch {
+	case atomic.LoadInt32(&c.pool.connected) != connected:
```
Comment (craiggwilson) [inline x/mongo/driver/topology/pool.go:88]: Can we add a method to the pool like we did for the connection to indicate whether it's closed?

`case c.pool.closed():`

File: x/mongo/driver/topology/cmap_prose_test.go [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+				assertConnectionCounts(t, pool, numConns, 0)
+
+				// Return all connections to the pool and assert that none were closed by put().
+				for i, c := range conns {
+					err := pool.put(c)
+					assert.Nil(t, err, "put error at index %d: %v", i, err)
+				}
+				assertConnectionCounts(t, pool, numConns, 0)
+
+				// Disconnenct the pool and assert that a closed event is published for each connection.
```
Comment (iwysiu) [inline x/mongo/driver/topology/cmap_prose_test.go:191]: Disconnect

File: x/mongo/driver/topology/cmap_prose_test.go [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+			}
+		}
+		assertConnectionCounts := func(t *testing.T, p *pool, numCreated, numClosed int) {
+			t.Helper()
+
+			assert.Equal(t, numCreated, len(created), "expected %d creation events, got %d", numCreated, len(created))
+			assert.Equal(t, numClosed, len(closed), "expected %d closed events, got %d", numClosed, len(closed))
+
+			netCount := numCreated - numClosed
+			assert.Equal(t, netCount, len(p.opened), "expected %d conenctions in openend map, got %d", netCount,
```
[the remark this answers is not in the document]
Comment (divjotarora) [reply x/mongo/driver/topology/cmap_prose_test.go]: done.

File: x/mongo/driver/topology/cmap_prose_test.go (hunk 2 above)
Comment (divjotarora) [reply x/mongo/driver/topology/cmap_prose_test.go]: done.

**Pull request merged**